client/app/components/queries/editor-components/databricks/useDatabricksSchema.js

```diff
@@ -31,22 +31,82 @@ export default function useDatabricksSchema(dataSource, options = null, onOption
   const [currentDatabaseName, setCurrentDatabaseName] = useState();
   const [schemas, setSchemas] = useState({});
   const [loadingSchema, setLoadingSchema] = useState(false);
+  const [refreshing, setRefreshing] = useState(false);
+
+  const setCurrentSchema = useCallback(
+    schema =>
+      setSchemas(currentSchemas => ({
+        ...currentSchemas,
+        [currentDatabaseName]: schema,
+      })),
+    [currentDatabaseName]
+  );
+
+  const currentDatabaseNameRef = useRef();
+  currentDatabaseNameRef.current = currentDatabaseName;
+  const loadTableColumns = useCallback(
+    tableName => {
+      // remove [databaseName.] from the tableName
+      DatabricksDataSource.getTableColumns(
+        dataSource,
+        currentDatabaseName,
+        tableName.substring(currentDatabaseName.length + 1)
+      ).then(columns => {
+        if (currentDatabaseNameRef.current === currentDatabaseName) {
+          setSchemas(currentSchemas => {
+            const schema = get(currentSchemas, currentDatabaseName, []);
+            const updatedSchema = map(schema, table => {
+              if (table.name === tableName) {
+                return { ...table, columns, loading: false };
+              }
+              return table;
+            });
+            return {
+              ...currentSchemas,
+              [currentDatabaseName]: updatedSchema,
+            };
+          });
+        }
+      });
+    },
+    [dataSource, currentDatabaseName]
+  );
 
   const schema = useMemo(() => get(schemas, currentDatabaseName, []), [schemas, currentDatabaseName]);
 
+  const refreshAll = useCallback(() => {
+    if (!refreshing) {
+      setRefreshing(true);
+      const getDatabasesPromise = getDatabases(dataSource, true).then(setDatabases);
+      const getSchemasPromise = getSchema(dataSource, currentDatabaseName, true).then(({ schema }) =>
+        setCurrentSchema(schema)
+      );
+
+      Promise.all([getSchemasPromise.catch(() => {}), getDatabasesPromise.catch(() => {})]).then(() =>
+        setRefreshing(false)
+      );
+    }
+  }, [dataSource, currentDatabaseName, setCurrentSchema, refreshing]);
+
   const schemasRef = useRef();
   schemasRef.current = schemas;
   useEffect(() => {
     let isCancelled = false;
     if (currentDatabaseName && !has(schemasRef.current, currentDatabaseName)) {
       setLoadingSchema(true);
       getSchema(dataSource, currentDatabaseName)
-        .then(data => {
+        .catch(() => Promise.resolve([]))
```

Comment:
Here an empty array is returned, but on next line an object is expected. Is it okay?

Reply:
Good catch, thanks! 🙂